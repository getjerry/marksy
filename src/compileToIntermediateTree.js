import marked from 'marked';
import he from 'he';

function createRenderer(tracker) {
  const renderer = new marked.Renderer();

  function populateInlineContent(content = '') {
    if (typeof content !== 'string') {
      return content;
    }
    const contentArray = content.split(/(\{\{.*?\}\})/);
    const extractedElements = contentArray
      .map(text => {
        const elementIdMatch = text.match(/\{\{(.*)\}\}/);
        if (elementIdMatch) {
          tracker.tree.splice(tracker.tree.indexOf(tracker.elements[elementIdMatch[1]]), 1);
          return tracker.elements[elementIdMatch[1]];
        }
        if (text !== '') {
          return he.decode(text);
        }

        return null;
      })
      .filter(one => one !== null && one !== undefined);

    return extractedElements;
  }

  function buildElement(tag, props = {}, children, addIdAsKey = true) {
    // eslint-disable-next-line no-plusplus, no-param-reassign
    const elementId = tracker.nextElementId++;
    let processedChildren = null;
    const processedProps = props;

    if (children) {
      processedChildren = Array.isArray(children)
        ? children.map(populateInlineContent).filter(one => one !== null && one !== undefined)
        : populateInlineContent(children);
    }
    if (addIdAsKey && processedProps) {
      processedProps.key = `md_${elementId}`;
    }

    // eslint-disable-next-line no-param-reassign
    tracker.elements[elementId] = {
      tag,
      ...(processedProps === null || processedProps === undefined ? {} : { props: processedProps }),
      ...(processedChildren === null || processedChildren === undefined
        ? {}
        : { children: processedChildren }),
    };
    // [tag, processedProps, processedChildren];
    tracker.tree.push(tracker.elements[elementId]);
    return `{{${elementId}}}`;
  }

  // Notice: need post process when render
  renderer.code = (code, language) => buildElement('code', { code, language }, null);

  renderer.html = html =>
    buildElement(
      'div',
      {
        dangerouslySetInnerHTML: {
          __html: html,
        },
      },
      null,
      true
    );

  renderer.paragraph = text => buildElement('p', null, text);

  renderer.blockquote = text => buildElement('blockquote', null, text);

  renderer.link = (href, title, text) => buildElement('a', { href, title }, text);

  renderer.br = () => buildElement('br');

  renderer.hr = () => buildElement('hr');

  renderer.strong = text => buildElement('strong', null, text);

  renderer.del = text => buildElement('del', null, text);

  renderer.em = text => buildElement('em', null, text);

  // Notice: need post process when render
  renderer.heading = (text, level) => buildElement('heading', { text, level }, null);

  renderer.list = (body, ordered) => buildElement(ordered ? 'ol' : 'ul', null, body);

  renderer.listitem = text => buildElement('li', null, text);

  renderer.table = (header, body) =>
    buildElement('table', null, [
      buildElement('thead', null, header),
      buildElement('tbody', null, body),
    ]);

  renderer.thead = content => buildElement('thead', null, content);

  renderer.tbody = content => buildElement('tbody', null, content);

  renderer.tablerow = content => buildElement('tr', null, content);

  renderer.tablecell = (content, flag) => {
    const tag = flag.header ? 'th' : 'td';
    return buildElement(tag, { className: flag.align ? `text-${flag.align}` : undefined }, content);
  };

  renderer.codespan = text => buildElement('codespan', null, text, 'codespan');

  renderer.image = (href, title, text) => buildElement('img', { src: href, alt: text });

  return renderer;
}

export function compileToIntermediateTree(content, options = {}, markedOptions = {}, context = {}) {
  const tracker = {
    tree: null,
    elements: null,
    nextElementId: null,
    toc: null,
    currentId: [],
  };
  const renderer = createRenderer(tracker, options, {});

  tracker.tree = [];
  tracker.elements = {};
  tracker.toc = [];
  tracker.nextElementId = 0;
  tracker.context = context;
  tracker.currentId = [];
  marked(content, { renderer, smartypants: true, ...markedOptions });

  return { tree: tracker.tree };
}
