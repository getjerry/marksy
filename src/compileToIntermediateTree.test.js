/* eslint-disable react/destructuring-assignment */
/* eslint-disable react/jsx-one-expression-per-line */
import React, { Component } from 'react';
import PropTypes from 'prop-types';

import hljs from 'highlight.js/lib/highlight';
import hljsJs from 'highlight.js/lib/languages/javascript';
import hljsXml from 'highlight.js/lib/languages/xml';

// eslint-disable-next-line
import { compileToIntermediateTree } from './compileToIntermediateTree';

hljs.registerLanguage('javascript', hljsJs);
hljs.registerLanguage('xml', hljsXml);

// eslint-disable-next-line
class TestComponent extends Component {
  render() {
    return <div>{this.props.children}</div>;
  }
}

TestComponent.propTypes = {
  children: PropTypes.node,
};

TestComponent.defaultProps = {
  children: null,
};

it('should be able to compile text', () => {
  const compiled = compileToIntermediateTree('hello');

  expect(compiled).toMatchSnapshot();
});

it('should be able to compile strong text', () => {
  const compiled = compileToIntermediateTree('hello **there**');

  expect(compiled).toMatchSnapshot();
});

it('should be able to compile italic text', () => {
  const compiled = compileToIntermediateTree('hello *there*');

  expect(compiled).toMatchSnapshot();
});

it('should be able to compile links', () => {
  const compiled = compileToIntermediateTree('[my link](http://example.com)');

  expect(compiled).toMatchSnapshot();
});

it('should be able to compile headers', () => {
  const compiled = compileToIntermediateTree(`
# header1
## header2
### header3
#### header4
  `);

  expect(compiled).toMatchSnapshot();
});

it('should handle same name nested headers', () => {
  const compiled = compileToIntermediateTree(`
# header1
## header2
# header3
## header2
  `);

  expect(compiled).toMatchSnapshot();
});

it('should be able to compile ordered list', () => {
  const compiled = compileToIntermediateTree(`
1. foo
2. bar
  `);

  expect(compiled).toMatchSnapshot();
});

it('should be able to compile list', () => {
  const compiled = compileToIntermediateTree(`
- foo
- bar
  `);

  expect(compiled).toMatchSnapshot();
});

it('should be able to compile tables', () => {
  const compiled = compileToIntermediateTree(`
| Tables        | Are           | Cool  |
| ------------- |:-------------:| -----:|
| col 3 is      | right-aligned | $1600 |
| col 2 is      | centered      |   $12 |
| zebra stripes | are neat      |    $1 |
  `);

  expect(compiled).toMatchSnapshot();
});

it('should be able to compile codespans', () => {
  const compiled = compileToIntermediateTree('install with `$ npm install`');

  expect(compiled).toMatchSnapshot();
});

it('should be able to compile image', () => {
  const compiled = compileToIntermediateTree('![test](http://some.com/image.png)');

  expect(compiled).toMatchSnapshot();
});

it('should be able to compile html', () => {
  const compiled = compileToIntermediateTree('<div>hello</div>');

  expect(compiled).toMatchSnapshot();
});

it('should be able to compile multiple html', () => {
  const compiled = compileToIntermediateTree(
    '<div>hello</div>\n<strong>there</strong>\n<em>world</em>'
  );

  expect(compiled).toMatchSnapshot();
});

it('should be able to compile components using H and marksy language', () => {
  const compiled = compileToIntermediateTree('```marksy\nh(Test)\n```');

  expect(compiled).toMatchSnapshot();
});

it('should be able to compile nested lists', () => {
  const compiled = compileToIntermediateTree(`
- Colors
    - Red
    - Blue
- Shape
  - Triangle
  - Rectangle
  `);

  expect(compiled).toMatchSnapshot();
});

it('should be able to combine in compilation', () => {
  const compiled = compileToIntermediateTree(`
# hey

- foo

-bar
  `);

  expect(compiled).toMatchSnapshot();
});

it('should produce TOC', () => {
  const compiled = compileToIntermediateTree(`
# foo

## bar

### baz
  `);

  expect(JSON.stringify(compiled.toc, null, 2)).toMatchSnapshot();
});

it('should allow injecting context to elements', () => {
  const compiled = compileToIntermediateTree(
    `
# foo
  `,
    {},
    {
      foo: 'bar',
    }
  );

  expect(compiled).toMatchSnapshot();
});

it('should be able to compile self-closing tag', () => {
  const compiled = compileToIntermediateTree(`
  ![test](http://some.com/image.png)
  
  <div><br /></div>
  
  <hr/>

  <input type='text' />
  `);

  expect(compiled).toMatchSnapshot();
});
