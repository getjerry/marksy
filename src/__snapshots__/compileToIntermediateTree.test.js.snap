// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should allow injecting context to elements 1`] = `
Object {
  "tree": Array [
    Object {
      "props": Object {
        "key": "md_0",
        "level": 1,
        "text": "foo",
      },
      "tag": "heading",
    },
  ],
}
`;

exports[`should be able to combine in compilation 1`] = `
Object {
  "tree": Array [
    Object {
      "props": Object {
        "key": "md_0",
        "level": 1,
        "text": "hey",
      },
      "tag": "heading",
    },
    Object {
      "children": Array [
        Object {
          "children": Array [
            "foo",
          ],
          "props": null,
          "tag": "li",
        },
      ],
      "props": null,
      "tag": "ul",
    },
    Object {
      "children": Array [
        "-bar",
      ],
      "props": null,
      "tag": "p",
    },
  ],
}
`;

exports[`should be able to compile codespans 1`] = `
Object {
  "tree": Array [
    Object {
      "children": Array [
        "install with ",
        Object {
          "children": Array [
            "$ npm install",
          ],
          "props": null,
          "tag": "codespan",
        },
      ],
      "props": null,
      "tag": "p",
    },
  ],
}
`;

exports[`should be able to compile components using H and marksy language 1`] = `
Object {
  "tree": Array [
    Object {
      "props": Object {
        "code": "h(Test)",
        "key": "md_0",
        "language": "marksy",
      },
      "tag": "code",
    },
  ],
}
`;

exports[`should be able to compile headers 1`] = `
Object {
  "tree": Array [
    Object {
      "props": Object {
        "key": "md_0",
        "level": 1,
        "text": "header1",
      },
      "tag": "heading",
    },
    Object {
      "props": Object {
        "key": "md_1",
        "level": 2,
        "text": "header2",
      },
      "tag": "heading",
    },
    Object {
      "props": Object {
        "key": "md_2",
        "level": 3,
        "text": "header3",
      },
      "tag": "heading",
    },
    Object {
      "props": Object {
        "key": "md_3",
        "level": 4,
        "text": "header4",
      },
      "tag": "heading",
    },
  ],
}
`;

exports[`should be able to compile html 1`] = `
Object {
  "tree": Array [
    Object {
      "props": Object {
        "dangerouslySetInnerHTML": Object {
          "__html": "<div>hello</div>",
        },
        "key": "md_0",
      },
      "tag": "div",
    },
  ],
}
`;

exports[`should be able to compile image 1`] = `
Object {
  "tree": Array [
    Object {
      "children": Array [
        Object {
          "props": Object {
            "alt": "test",
            "key": "md_0",
            "src": "http://some.com/image.png",
          },
          "tag": "img",
        },
      ],
      "props": null,
      "tag": "p",
    },
  ],
}
`;

exports[`should be able to compile italic text 1`] = `
Object {
  "tree": Array [
    Object {
      "children": Array [
        "hello ",
        Object {
          "children": Array [
            "there",
          ],
          "props": null,
          "tag": "em",
        },
      ],
      "props": null,
      "tag": "p",
    },
  ],
}
`;

exports[`should be able to compile links 1`] = `
Object {
  "tree": Array [
    Object {
      "children": Array [
        Object {
          "children": Array [
            "my link",
          ],
          "props": Object {
            "href": "http://example.com",
            "key": "md_0",
            "title": null,
          },
          "tag": "a",
        },
      ],
      "props": null,
      "tag": "p",
    },
  ],
}
`;

exports[`should be able to compile list 1`] = `
Object {
  "tree": Array [
    Object {
      "children": Array [
        Object {
          "children": Array [
            "foo",
          ],
          "props": null,
          "tag": "li",
        },
        Object {
          "children": Array [
            "bar",
          ],
          "props": null,
          "tag": "li",
        },
      ],
      "props": null,
      "tag": "ul",
    },
  ],
}
`;

exports[`should be able to compile multiple html 1`] = `
Object {
  "tree": Array [
    Object {
      "props": Object {
        "dangerouslySetInnerHTML": Object {
          "__html": "<div>hello</div>
<strong>there</strong>
<em>world</em>",
        },
        "key": "md_0",
      },
      "tag": "div",
    },
  ],
}
`;

exports[`should be able to compile nested lists 1`] = `
Object {
  "tree": Array [
    Object {
      "children": Array [
        Object {
          "children": Array [
            "Colors",
            Object {
              "children": Array [
                Object {
                  "children": Array [
                    "Red",
                  ],
                  "props": null,
                  "tag": "li",
                },
                Object {
                  "children": Array [
                    "Blue",
                  ],
                  "props": null,
                  "tag": "li",
                },
              ],
              "props": null,
              "tag": "ul",
            },
          ],
          "props": null,
          "tag": "li",
        },
        Object {
          "children": Array [
            "Shape",
            Object {
              "children": Array [
                Object {
                  "children": Array [
                    "Triangle",
                  ],
                  "props": null,
                  "tag": "li",
                },
                Object {
                  "children": Array [
                    "Rectangle",
                  ],
                  "props": null,
                  "tag": "li",
                },
              ],
              "props": null,
              "tag": "ul",
            },
          ],
          "props": null,
          "tag": "li",
        },
      ],
      "props": null,
      "tag": "ul",
    },
  ],
}
`;

exports[`should be able to compile ordered list 1`] = `
Object {
  "tree": Array [
    Object {
      "children": Array [
        Object {
          "children": Array [
            "foo",
          ],
          "props": null,
          "tag": "li",
        },
        Object {
          "children": Array [
            "bar",
          ],
          "props": null,
          "tag": "li",
        },
      ],
      "props": null,
      "tag": "ol",
    },
  ],
}
`;

exports[`should be able to compile self-closing tag 1`] = `
Object {
  "tree": Array [
    Object {
      "children": Array [
        "  ",
        Object {
          "props": Object {
            "alt": "test",
            "key": "md_0",
            "src": "http://some.com/image.png",
          },
          "tag": "img",
        },
      ],
      "props": null,
      "tag": "p",
    },
    Object {
      "props": Object {
        "dangerouslySetInnerHTML": Object {
          "__html": "  <div><br /></div>

",
        },
        "key": "md_2",
      },
      "tag": "div",
    },
    Object {
      "props": Object {
        "dangerouslySetInnerHTML": Object {
          "__html": "  <hr/>

",
        },
        "key": "md_3",
      },
      "tag": "div",
    },
    Object {
      "props": Object {
        "dangerouslySetInnerHTML": Object {
          "__html": "  <input type='text' />
",
        },
        "key": "md_4",
      },
      "tag": "div",
    },
  ],
}
`;

exports[`should be able to compile strong text 1`] = `
Object {
  "tree": Array [
    Object {
      "children": Array [
        "hello ",
        Object {
          "children": Array [
            "there",
          ],
          "props": null,
          "tag": "strong",
        },
      ],
      "props": null,
      "tag": "p",
    },
  ],
}
`;

exports[`should be able to compile tables 1`] = `
Object {
  "tree": Array [
    Object {
      "children": Array [
        Array [
          Object {
            "children": Array [
              Object {
                "children": Array [
                  Object {
                    "children": Array [
                      "Tables",
                    ],
                    "props": Object {
                      "className": undefined,
                      "key": "md_0",
                    },
                    "tag": "th",
                  },
                  Object {
                    "children": Array [
                      "Are",
                    ],
                    "props": Object {
                      "className": "text-center",
                      "key": "md_1",
                    },
                    "tag": "th",
                  },
                  Object {
                    "children": Array [
                      "Cool",
                    ],
                    "props": Object {
                      "className": "text-right",
                      "key": "md_2",
                    },
                    "tag": "th",
                  },
                ],
                "props": null,
                "tag": "tr",
              },
            ],
            "props": null,
            "tag": "thead",
          },
        ],
        Array [
          Object {
            "children": Array [
              Object {
                "children": Array [
                  Object {
                    "children": Array [
                      "col 3 is",
                    ],
                    "props": Object {
                      "className": undefined,
                      "key": "md_4",
                    },
                    "tag": "td",
                  },
                  Object {
                    "children": Array [
                      "right-aligned",
                    ],
                    "props": Object {
                      "className": "text-center",
                      "key": "md_5",
                    },
                    "tag": "td",
                  },
                  Object {
                    "children": Array [
                      "$1600",
                    ],
                    "props": Object {
                      "className": "text-right",
                      "key": "md_6",
                    },
                    "tag": "td",
                  },
                ],
                "props": null,
                "tag": "tr",
              },
              Object {
                "children": Array [
                  Object {
                    "children": Array [
                      "col 2 is",
                    ],
                    "props": Object {
                      "className": undefined,
                      "key": "md_8",
                    },
                    "tag": "td",
                  },
                  Object {
                    "children": Array [
                      "centered",
                    ],
                    "props": Object {
                      "className": "text-center",
                      "key": "md_9",
                    },
                    "tag": "td",
                  },
                  Object {
                    "children": Array [
                      "$12",
                    ],
                    "props": Object {
                      "className": "text-right",
                      "key": "md_10",
                    },
                    "tag": "td",
                  },
                ],
                "props": null,
                "tag": "tr",
              },
              Object {
                "children": Array [
                  Object {
                    "children": Array [
                      "zebra stripes",
                    ],
                    "props": Object {
                      "className": undefined,
                      "key": "md_12",
                    },
                    "tag": "td",
                  },
                  Object {
                    "children": Array [
                      "are neat",
                    ],
                    "props": Object {
                      "className": "text-center",
                      "key": "md_13",
                    },
                    "tag": "td",
                  },
                  Object {
                    "children": Array [
                      "$1",
                    ],
                    "props": Object {
                      "className": "text-right",
                      "key": "md_14",
                    },
                    "tag": "td",
                  },
                ],
                "props": null,
                "tag": "tr",
              },
            ],
            "props": null,
            "tag": "tbody",
          },
        ],
      ],
      "props": null,
      "tag": "table",
    },
  ],
}
`;

exports[`should be able to compile text 1`] = `
Object {
  "tree": Array [
    Object {
      "children": Array [
        "hello",
      ],
      "props": null,
      "tag": "p",
    },
  ],
}
`;

exports[`should handle same name nested headers 1`] = `
Object {
  "tree": Array [
    Object {
      "props": Object {
        "key": "md_0",
        "level": 1,
        "text": "header1",
      },
      "tag": "heading",
    },
    Object {
      "props": Object {
        "key": "md_1",
        "level": 2,
        "text": "header2",
      },
      "tag": "heading",
    },
    Object {
      "props": Object {
        "key": "md_2",
        "level": 1,
        "text": "header3",
      },
      "tag": "heading",
    },
    Object {
      "props": Object {
        "key": "md_3",
        "level": 2,
        "text": "header2",
      },
      "tag": "heading",
    },
  ],
}
`;

exports[`should produce TOC 1`] = `undefined`;
